### PR TITLE
Edit root.go for binding ENV variables

### LINKED ISSUE #230: Bind ENV vars and Flags to config
We are using [Cobra](https://cobra.dev/) and [Viper](https://github.com/spf13/viper) for CLI and Config handles.

The idea is to allow the users to provide the configuration of the node using also the CLI commands and ENV vars

### PR DESCRIPTION
**Pull request by smishy05**
Resolves #230 

## Changes:
- Adds cli commands for updating the config file using the flags and environment variables

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [ ] Yes
- [x] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [ ] No

**Comments about testing , should you have some** (optional)

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: cmd/juno/cli/root.go [hunk 1]
```diff
@@ -56,6 +57,124 @@ var (
...
 			}(sig)
 
+			// Check if the user has provided any value for configurations
+			// The order of preference
+			// 1. Values provided with flags
+			// 2. Values provided using environment variables
+			// 3. Values from config file (the default)
+			// Note that the config file isn't modified by these flags.
+			// The values provided to the flags or environment variables are
+			// used for that particular run only
```
Comment (D-DePablos) [inline cmd/juno/cli/root.go:67]: good!

Comment (smishy05) [reply cmd/juno/cli/root.go:67]: thanks!

File: cmd/juno/cli/root.go [hunk 2]
```diff
@@ -56,6 +57,124 @@ var (
...
+
+			// This variable keeps a count of the number of flags
+			// If it is zero and the user asks for updating the config file,
```
[the remark this answers is not in the document]
Comment (smishy05) [reply cmd/juno/cli/root.go]: Done!

**Review by D-DePablos — DISMISSED** [review]

File: cmd/juno/cli/root.go [hunk 3]
```diff
@@ -138,11 +199,62 @@ var (
 	}
 )
 
+func handleConfig(configParam, flag, envVarName string, t int) error {
```
Comment (D-DePablos) [inline cmd/juno/cli/root.go:202]: what is t?

Comment (smishy05) [reply cmd/juno/cli/root.go:207]: Added relevant comments for the arguments.

File: cmd/juno/cli/root.go [hunk 4]
```diff
@@ -153,6 +270,27 @@ func init() {
 		"config file (default is %s)", filepath.Join(config.Dir, "juno.yaml")))
 	rootCmd.PersistentFlags().StringVar(&dataDir, "dataDir", "", fmt.Sprintf(
 		"data path (default is %s)", config.DataDir))
+	// RPC
+	rootCmd.PersistentFlags().StringP("rpcport", "p", os.Getenv("RPCPORT"), "Set the RPC Port")
```
Comment (adriantpaez) [inline cmd/juno/cli/root.go:274]: Any issue using `Flags()` instead `PersistentFlags()`? I think is better to use `Flags()` if is possible

**Review by adriantpaez — CHANGES_REQUESTED** [review]
I think more separation between Cobra and Viper is needed, when I read the code I can't know properly if the line is about a Cobra flag or a Viper config variable. Maybe a function like `setupViper(cmd *cobra.Command)` and putting all the Viper things there can make the code clear.

File: cmd/juno/cli/root.go [hunk 5]
```diff
@@ -56,6 +57,56 @@ var (
...
+			handleConfig("rest.prefix", flag, "RESTPREFIX", 2)
+
+			erru := viper.Unmarshal(&config.Runtime)
+			errpkg.CheckFatal(erru, "Unable to unmarshal runtime config instance.")
+
+			flag, _ = cmd.PersistentFlags().GetString("updateconfigfile")
+
+			if flag != "" {
+				updateConfigFile(cfgFile)
+			}
```
[the remark this answers is not in the document]
Comment (smishy05) [reply cmd/juno/cli/root.go:108]: Not exactly. It's only updated only if the user sets the flag `U` to true. Juno runs with the flag provided for one instance of the run. But, let's say the user wants to have that value for every run. For this reason, I added another flag. Does this make sense now?

**Comment by smishy05** [comment]
> 

I think the function `handleConfig` does the job of updating the viper config parameters. Does that solve the purpose or should I think of separating Viper and Cobra more?

File: cmd/juno/cli/root.go [hunk 6]
```diff
@@ -56,6 +57,56 @@ var (
...
+			// 1. Values provided with flags
+			// 2. Values provided using environment variables
+			// 3. Values from config file (the default)
+			// Note that the config file isn't modified by these flags.
+			// The values provided to the flags or environment variables are
+			// used for that particular run only
+			// To update the config file, the user needs to make the
+			// -U flag true
+
+			flag, _ := cmd.PersistentFlags().GetString("rpcenable")
```
[the remark this answers is not in the document]
Comment (smishy05) [reply cmd/juno/cli/root.go:71]: Same as the reply in `int`

File: cmd/juno/cli/root.go [hunk 7]
```diff
@@ -56,6 +57,56 @@ var (
...
+			handleConfig("rpc.enabled", flag, "RPCENABLE", 0)
+			flag, _ = cmd.PersistentFlags().GetString("rpcport")
```
[the remark this answers is not in the document]
Comment (smishy05) [reply cmd/juno/cli/root.go:73]: This is so as to do everything using only one function (`handleConfig`). If I have different datatypes, then, I would need to have different function definition for each type. Which one should I prefer then - `GetString` or `GetInt`?

File: cmd/juno/cli/root.go (hunk 4 above)
Comment (smishy05) [reply cmd/juno/cli/root.go]: Done.

File: cmd/juno/cli/root.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (smishy05) [reply cmd/juno/cli/root.go]: Done. I have converted them to their corresponding datatypes

File: cmd/juno/cli/root.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (smishy05) [reply cmd/juno/cli/root.go]: Done. I have converted them to their corresponding datatypes

File: cmd/juno/cli/root.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (smishy05) [reply cmd/juno/cli/root.go]: done

**Comment by joshklop** [comment]
Closing in favor of #295

**Pull request closed without merging**